draft-ietf-httpbis-origin-frame.md

```diff
@@ -70,80 +75,183 @@ The ORIGIN HTTP/2 frame ({{!RFC7540}}, Section 4) allows a server to indicate wh
-4. For each Origin field `origin_raw` in the frame payload:
-   1. Parse `origin_raw` as an ASCII serialization of an origin ({{!RFC6454}}, Section 6.2) and let the result be `parsed_origin`.
-   2. If the REMOVE flag is set, remove any member of the Origin Set that is the same as `parsed_origin` (as per {{!RFC6454}}, Section 5), and continue to the next `parsed_origin`.
-   3. Otherwise, add `parsed_origin` to the Origin Set.
+# IANA Considerations
+
+This specification adds an entry to the "HTTP/2 Frame Type" registry.
+
+* Frame Type: ORIGIN
+* Code: 0xb
+* Specification: [this document]
 
 
 # Security Considerations
 
 Clients that blindly trust the ORIGIN frame's contents will be vulnerable to a large number of
-attacks; hence the reinforcement that this specification does not relax the requirement for server
-authority in {{?RFC7540}}, Section 10.1.
+attacks. See {{authority}} for mitigations.
+
+Relaxing the requirement to consult DNS when determining authority for an origin means that an
+attacker who possesses a valid certificate no longer needs to be on-path to redirect traffic to
+them; instead of modifying DNS, they need only convince the user to visit another Web site, in
+order to coalesce connections to the target onto their existing connection.
 
 --- back
+
+
+# Non-Normative Processing Algorithm {#algo}
+
+The following algorithm illustrates how a client could handle received ORIGIN frames:
+
+1. If the client is configured to use a proxy for the connection, ignore the frame and stop processing.
+2. If the connection is not identified with the "h2" protocol identifier or another protocol that has explicitly opted into this specification, ignore the frame and stop processing.
+3. If the frame occurs upon any stream except stream 0, ignore the frame and stop processing.
+4. If any of the flags 0x1, 0x2, 0x4 or 0x8 are set, ignore the frame and stop processing.
+5. If no previous ORIGIN frame on the connection has reached this step, initialise the Origin Set as per {{set}}.
+6. For each Origin field `origin_raw` in the frame payload:
+   1. Parse `origin_raw` as an ASCII serialization of an origin ({{!RFC6454}}, Section 6.2) and let the result be `parsed_origin`. If parsing fails, skip to the next `origin_raw`.
+   2. If the `scheme` of `parsed_origin` is not "https", skip to the next `origin_raw`.
+   3. If the certificate presented by the server is not valid for the `host` of `parsed_origin` (see below), skip to the next `origin_raw`.
+   4. If the `port` of `parsed_origin` does not match the connection's remote port, skip to the next `origin_raw`.
+   5. Add `parsed_origin` to the Origin Set.
+
+The certificate presented by the server is valid for a host if it passes the checks that the client
+would perform when forming a new TLS connection to the origin. This includes verifying that the
+host matches a `dNSName` value from the certificate `subjectAltName` field (using the wildcard rules
```

Comment:
Is correct; see https://tools.ietf.org/html/rfc5280#section-4.2.1.6 (but will add a ref)